### PR TITLE
add ut

### PR DESCRIPTION
**Pull request by galaxydi**
1. add ut
2. refine AliLogAPI interface

Now the test coverage is 66%, already cover the main point

### CODE REVIEW

File: daemon/logger/alilogs/alilogs.go [hunk 1]
```diff
@@ -170,42 +152,39 @@ var newTicker = func(freq time.Duration) *time.Ticker {
...
 		case <-timer.C:
-			ls.publishLogs(&logGroup)
+			ls.publishLogs()
 			logrus.WithFields(logrus.Fields{
-				"endpoint":             aliLogClient.Endpoint,
-				"project":              aliLogClient.ProjectName,
-				"logstore":             aliLogClient.LogstoreName,
-				"published log number": len(logGroup.Logs),
-				"published log size":   logGroup.Size(),
+				"time trigger":         "send data",
```
Comment (wanghq) [inline daemon/logger/alilogs/alilogs.go]: I can see that the publish could be triggered by 3 kind of events, time, size, and end of file. Let's pick up one key and different values. e.g.
```
trigger: time|size|eof
```

File: daemon/logger/alilogs/alilogs.go [hunk 2]
```diff
@@ -170,42 +152,39 @@ var newTicker = func(freq time.Duration) *time.Ticker {
...
+				"endpoint":             ls.endpoint,
+				"project":              ls.projectName,
+				"logstore":             ls.logstoreName,
+				"published log number": len(ls.logGroup.Logs),
+				"published log size":   ls.logGroup.Size(),
```
Comment (wanghq) [inline daemon/logger/alilogs/alilogs.go]: published log number -> logcount
published log size - > logsize

File: daemon/logger/alilogs/alilogs.go [hunk 3]
```diff
@@ -170,42 +152,39 @@ var newTicker = func(freq time.Duration) *time.Ticker {
...
 			logrus.WithFields(logrus.Fields{
-				"endpoint":             aliLogClient.Endpoint,
-				"project":              aliLogClient.ProjectName,
-				"logstore":             aliLogClient.LogstoreName,
-				"published log number": len(logGroup.Logs),
-				"published log size":   logGroup.Size(),
+				"time trigger":         "send data",
+				"endpoint":             ls.endpoint,
+				"project":              ls.projectName,
+				"logstore":             ls.logstoreName,
```
Comment (wanghq) [inline daemon/logger/alilogs/alilogs.go]: if you want to save few lines, you can create a common log entry and use it in 3 places. 
```
le := logrus.WithFields(logrus.Fields{
	"endpoint":ls.endpoint,
	"project":ls.projectName,
	"logstore":ls.logstoreName,
})
// ...
le.WithFields(logrus.Fields{
	"trigger": "time",
	"logcount": 234,
	"logsize": 123,
}).Debug("")
```

File: daemon/logger/alilogs/alilogs.go [hunk 4]
```diff
@@ -221,38 +200,38 @@ func (ls *logStream) collectLogs() {
...
 				"errorMessage": serviceErr.Message,
-				"endpoint":     aliLogClient.Endpoint,
-				"project":      aliLogClient.ProjectName,
-				"logstore":     aliLogClient.LogstoreName,
+				"endpoint":     ls.endpoint,
+				"project":      ls.projectName,
+				"logstore":     ls.logstoreName,
 			}).Error("PutLogs occurs sls error")
 		} else {
 			logrus.Error(err)
```
Comment (wanghq) [inline daemon/logger/alilogs/alilogs.go]: Can you add the context fields (endpoint/project/logstore) to this log? Also, log this as PutLogs error, e.g. `logrus.Error("PutLogs error: %s", err)`

File: daemon/logger/alilogs/alilogs_test.go [hunk 5]
```diff
@@ -0,0 +1,316 @@
...
+	})
+	stream.Log(&logger.Message{
+		Line:      []byte("this is test log 2"),
+		Timestamp: time.Time{},
+	})
+	stream.Log(&logger.Message{
+		Line:      []byte("this is test log 3"),
+		Timestamp: time.Time{},
+	})
+	time.Sleep(batchPublishFrequency)
```
Comment (wanghq) [inline daemon/logger/alilogs/alilogs_test.go]: Can you check how this [test](https://github.com/docker/docker/blob/master/daemon/logger/awslogs/cloudwatchlogs_test.go#L387-L405) works by mocking the newTicker? So you don't need to sleep.

File: daemon/logger/alilogs/mock_client_test.go [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+		Topic:   "",
+		Logs:    []*sls.Log{},
+	}
+	return client
+}
+
+// PutLogs provides a mock function with given fields: _a0
+func (client *MockClient) PutLogs(lg *sls.LogGroup) error {
+	switch client.ErrType {
+	case 0:
```
Comment (wanghq) [inline daemon/logger/alilogs/mock_client_test.go]: How about this so you don't need to reference 0, 1, 2 in the tests.
```
type errorType int

const (
	NoError errorType = iota
	InternalError
	UnknownError
)
```

**Comment by galaxydi** [comment]
Good tips, fix all your comments

**Pull request merged**